File: 727-minimum-window-subsequence/727-minimum-window-subsequence.cpp
class Solution {
public:
    string minWindow(string s1, string s2) {
        string ans = "";
        int n = s1.size(), m = s2.size();
        int mini = n+1;
        int j = 0, end = 0;
        for(int i=0;i<n;i++){
            if(s1[i]==s2[j]) j++;
            if(j==m){
                end = i;
                j--;
                while(j>=0){
                    if(s1[i]==s2[j]){
                        j--;
                    }
                    i--;
                }
                i++, j++;
                if(end - i + 1 < mini){
                    mini = end - i + 1;
                    ans = s1.substr(i, end-i+1);
                }
            }
        }
        return ans;
    }
};

File: 727-minimum-window-subsequence/727-minimum-window-subsequence.java
class Solution {
    public String minWindow(String S, String T) {
        String window = "";
        int j = 0, min = S.length() + 1;
        for (int i = 0; i < S.length(); i++) {
            if (S.charAt(i) == T.charAt(j)) {
                j++;
                if (j == T.length()) {
                    int end = i + 1;
                    j--;
                    while (j >= 0) {
                        if (S.charAt(i) == T.charAt(j)) j--;
                        i--;
                    }
                    j++;
                    i++;
                    if (end - i < min) {
                        min = end - i;
                        window = S.substring(i, end);
                    }
                }
            }
        }
        return window;
    }
}
